core/src/main/java/org/elasticsearch/client/transport/TransportClientNodesService.java

```diff
@@ -360,9 +360,10 @@ protected void doSample() {
                     try {
                         // its a listed node, light connect to it...
                         logger.trace("connecting to listed node (light) [{}]", listedNode);
-                        transportService.connectToNodeLight(listedNode);
+                        transportService.connectToNodeLight(listedNode, pingTimeout, !ignoreClusterName);
```

Comment:
OK - I did the digging. It is the responsibility of the `nodesSampler` to build the `DiscoveryNode`s that everything else uses. They (there are two implementations) use these prototypical DiscoveryNodes to do their job. They have a neat hack where they assume that the node that they are talking to is always the minimum compatible version. I haven't verified that the other side (the node's themselves) do this but I suspect they do. I haven't verified if the nodes have this hack when discovering each other but they probably do.

Anyway, it makes me think that the liveness check is actually being used for everything that we want to use the handshake for already. But only in the transport client?